Fixes parser for the top level tokens

abhikuhikar opened the pull request:
Issue #363 

Fixes parser by traversing and validating the parse tree to ensure that the top level tokens are not present at the level(s) below. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

alancai98 commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1246,6 +1246,16 @@ class ParserErrorsTest : TestBase() {
+        mapOf(
+            Property.LINE_NUMBER to 1L,
+            Property.COLUMN_NUMBER to 35L,
+            Property.TOKEN_TYPE to TokenType.EOF,
+            Property.TOKEN_VALUE to ion.newSymbol("EOF")))
```
Shouldn't the unexpected parse term be `SET` here and not `EOF`?

abhikuhikar replied:
You are right. Thanks. I have made changes in the next revision.

alancai98 reviewed (changes requested):
The approach lgtm overall. Could add more top-level token checks to `EXEC` and other DML/DDL ops. Also may need more tests.

abhikuhikar commented on lang/test/org/partiql/lang/syntax/SqlParserTest.kt:
```diff
@@ -2636,6 +2636,32 @@ class SqlParserTest : SqlParserTestBase() {
     )
 
+    @Test
+    fun fromWhereSetDml() = assertExpression(
```
Ohh yes. I missed that. Thanks. I will remove it.

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2259,6 +2259,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
+    private fun ParseNode.isTopLevelType() = type in listOf(UPDATE, DELETE, INSERT)
+
+    /**
+     * Validates tree to make sure that the top level tokens are not found below the top level
```
Since it is a tree representation, "below" makes sense IMO.

abhikuhikar commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1246,6 +1246,16 @@ class ParserErrorsTest : TestBase() {
             Property.TOKEN_TYPE to TokenType.EOF,
             Property.TOKEN_VALUE to ion.newSymbol("EOF")))
 
+    @Test
```
Thanks. Added tests for other top level tokens.

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2259,6 +2259,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
+        // Note that for DML operations, top level parse node is of type 'FROM'. Hence the check level > 1
+        if (level > 1 && node.isTopLevelType()) {
+            node.remaining.err("Type ${node.type} only expected at the top level", PARSE_UNEXPECTED_TERM)
+        }
```
I think the logic remains the same for the top level tokens even if it is DML or not.

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2259,6 +2259,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
+    /**
+     * Checks if the [ParseNode] requires to be only at top level. DML keyword tokens, for example, requires to be top level.
+     */
+    private fun ParseNode.isTopLevelType() = type in listOf(UPDATE, DELETE, INSERT)
```
Added.

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2259,6 +2259,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
 
+    /**
+     * Checks if the [ParseNode] requires to be only at top level. DML keyword tokens, for example, requires to be top level.
+     */
```
Added `EXEC`.

alancai98 replied:
Sorry, I may have put too much in the first comment, but will [`DDL` keywords](https://github.com/partiql/partiql-lang-kotlin/blob/master/lang/resources/org/partiql/type-domains/partiql.ion#L234-L250) also be added?

abhikuhikar replied:
We can always add token types to the top level types list.

alancai98 commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1623,10 +1673,10 @@ class ParserErrorsTest : TestBase() {
-            Property.TOKEN_VALUE to ion.newSymbol("exec")))
+            Property.COLUMN_NUMBER to 21L,
+            Property.TOKEN_TYPE to TokenType.IDENTIFIER,
+            Property.TOKEN_VALUE to ion.newSymbol("undrop")))
```
Shouldn't the unexpected term be `EXEC` here?

abhikuhikar replied:
It should be. But the associated token with the `EXEC` ParseNode is the name of the stored procedure. Hence, the issue. Ti get around with this, I think we can pass the name of the stored procedure in the arguments along with the other `arglist` or some such. But that's worthy of a separate issue.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2250,27 @@ class SqlParser(private val ion: IonSystem) : Parser {
-     *
-     * Currently only checks for `EXEC`. DDL and DML along with corresponding error codes to be added in the future
-     * (https://github.com/partiql/partiql-lang-kotlin/issues/354).
+     * Checks if the [ParseNode] requires to be only at top level. DML keyword tokens, for example, requires to be top level.
```
super nit: at least for me the kdoc wording is confusing. From the implementation, it checks the ParseNode's parse type is one of the top level types right? The first sentence could be something like:
```kotlin
/**
 * Returns true if and only if this [ParseType] is one of the top level types...
 */
```
For the second sentence, it might be helpful to also mention that DML can have `FROM` as the top-level node.

abhikuhikar replied:
I will update it. Thanks.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2250,27 @@ class SqlParser(private val ion: IonSystem) : Parser {
+     * Validates tree to make sure that the top level tokens are not found below the top level
+     */
+    private fun validateTopLevelNodes(node: ParseNode, level: Int) {
+        // Note that for DML operations, top level parse node is of type 'FROM'. Hence the check level > 1
```
super nit: the top level parse node **could** be of type `FROM` (not necessary for all DML ops)

abhikuhikar replied:
It is for all DML ops if I am not wrong. I have updated the logic to take this into consideration.

alancai98 replied:
From the [partiql.ion](https://github.com/partiql/partiql-lang-kotlin/blob/master/lang/resources/org/partiql/type-domains/partiql.ion#L18-L26) grammar for DML ops, the `FROM` source is optional.

alancai98 reviewed (commented):
Just a few minor nits for the comments/kdoc I missed from the first review. I probably overloaded [this comment](https://github.com/partiql/partiql-lang-kotlin/pull/369#discussion_r584919948) from the first review, but will DDL ops also be included?

abhikuhikar commented:
- Updated the logic to check for DDL commands as well. 
- Changed `ParseType` to include properties `isTopLevelType` and `isDml`.
- Added tests to cover DDL and DML top level token types.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2253,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
+    private fun validateTopLevelNodes(node: ParseNode, level: Int) {
+        // Note that for DML operations, top level parse node is of type 'FROM'. Hence the check level > 1
+        if (node.type.isTopLevelType && ((node.type.isDml && level > 1) || (!node.type.isDml && level > 0))) {
+            node.token?.err("Type ${node.type} only expected at the top level", PARSE_UNEXPECTED_TERM)
+                ?: throw ParserException("Type ${node.type} only expected at the top level", PARSE_UNEXPECTED_TERM, PropertyValueMap())
```
nit: I think the word "Type" shouldn't be included in the error message. Maybe it could be replaced with "Keyword"?

abhikuhikar replied:
So the `parseType` is different from the keyword or token, fortunately or unfortunately. I guess we can remove the error message - `type only expected at top level` altogether and just stay with the `unexpected term found` error  message.

abhikuhikar replied:
Changed the error msg.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2253,22 @@ class SqlParser(private val ion: IonSystem) : Parser {
-            }
+    private fun validateTopLevelNodes(node: ParseNode, level: Int) {
+        // Note that for DML operations, top level parse node is of type 'FROM'. Hence the check level > 1
+        if (node.type.isTopLevelType && ((node.type.isDml && level > 1) || (!node.type.isDml && level > 0))) {
```
nit: This can be a little hard to understand with so many logic symbols. Can this be simplified or broken into multiple if-statements?

abhikuhikar replied:
It will make the code redundant. I also thought the conditional statement is complicated initially  but on the second glance it is not that complicated. 

```
if (node.type.isTopLevelType && level > 0) {
    if (node.type.isDml) {
        if (level > 1) {
            // throw error
        }
    } else {
        // throw error
    }
}
```

abhikuhikar replied:
Changed as mentioned above.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -55,7 +55,10 @@ class SqlParser(private val ion: IonSystem) : Parser {
     }
 
-    internal enum class ParseType(val isJoin: Boolean = false) {
+    /** If the property [isTopLevelType] is true, then the parse node of this ParseType will only be valid at top level in the query.
+     *  For example, EXEC, DDL and DML keywords can only be used at the top level in the query.
+     */
+    internal enum class ParseType(val isJoin: Boolean = false, val isTopLevelType: Boolean = false, val isDml: Boolean = false) {
```
How is the `isDml` parameter defined? `SET` is a DML op, yet `ParseType.SET` `isDml` is `false`.

abhikuhikar replied:
This is a legit observation. However, the `SET` `parseType` is not used anywhere. We can actually get rid of it.

abhikuhikar replied:
Changed the property `isDml` to `true` for `SET`.

alancai98 commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1246,6 +1246,96 @@ class ParserErrorsTest : TestBase() {
 
+    @Test
+    fun updateWithNestedSet() = checkInputThrowingParserException(
+        "UPDATE test SET x = SET test.y = 6",
```
So when I'm testing this query in the REPL, it gives the following:

```
PartiQL> UPDATE test SET x = SET test.y = 6;
org.partiql.lang.syntax.ParserException: Type UPDATE only expected at the top level
	Parser Error: at line 1, column 21: unexpected term found, KEYWORD : set
```
Shouldn't the 2nd line say `SET` and not `UPDATE`? I haven't had a chance to test w/ the other top-level tokens.

abhikuhikar replied:
The associated parse type is `UPDATE`.
I will change this to - `"Keyword ${node.token} only expected at the top level"`.
It makes more sense I guess and would be less confusing.

abhikuhikar commented:
Came across interesting edge case where the DML tokens are at top level. Updated the PR to handle these edge cases.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2272,6 +2295,9 @@ class SqlParser(private val ion: IonSystem) : Parser {
             }
         }
+
+        validateTopLevelNodes(node, 0, 0)
```
nit: could add names to call args

```kotlin
validateTopLevelNodes(node = node, level = 0, topLevelTokens = 0)
```

abhikuhikar replied:
Added.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2253,39 @@ class SqlParser(private val ion: IonSystem) : Parser {
+                node.throwTopLevelParserError()
             }
         }
+        node.children.map { validateTopLevelNodes(it, level + 1, topTokens) }
```
nit: could add names to call args:

```kotlin
node.children.map { validateTopLevelNodes(node = it, level = level + 1, topLevelTokens = topTokens) }
```

abhikuhikar replied:
Added.

alancai98 commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1246,6 +1246,116 @@ class ParserErrorsTest : TestBase() {
+    @Test
+    fun nestedRemove() = checkInputThrowingParserException(
+        "REMOVE REMOVE y",
+        ErrorCode.PARSE_INVALID_PATH_COMPONENT,
```
Is there a reason this error is not `PARSE_UNEXPECTED_TERM`?

abhikuhikar replied:
This is handled actively by the parser while parsing the `REMOVE` keyword.

alancai98 reviewed (commented):
Just have a few minor nits and questions. The rest lgtm

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2253,39 @@ class SqlParser(private val ion: IonSystem) : Parser {
     }
 
+    private fun ParseNode.throwTopLevelParserError(): Nothing =
+        token?.err("Keyword $token only expected at the top level in the query", PARSE_UNEXPECTED_TERM)
```
Makes sense. Thanks. Changed it.

abhikuhikar commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2244,25 +2253,39 @@ class SqlParser(private val ion: IonSystem) : Parser {
 
+    private fun ParseNode.throwTopLevelParserError(): Nothing =
+        token?.err("Keyword $token only expected at the top level in the query", PARSE_UNEXPECTED_TERM)
+        ?: throw ParserException("Keyword $token only expected at the top level in the query", PARSE_UNEXPECTED_TERM, PropertyValueMap())
+
     /**
-     * Checks that the given [Token] list does not have any top-level tokens outside of the top level query. Throws
-     * an error if so.
-     *
-     * Currently only checks for `EXEC`. DDL and DML along with corresponding error codes to be added in the future
-     * (https://github.com/partiql/partiql-lang-kotlin/issues/354).
+     * Validates tree to make sure that the top level tokens are not found below the top level
      */
-    private fun List<Token>.checkUnexpectedTopLevelToken() {
-        for ((index, token) in this.withIndex()) {
-            if (token.keywordText == "exec" && index != 0) {
-                token.err("EXEC call found at unexpected location", PARSE_EXEC_AT_UNEXPECTED_LOCATION)
+    private fun validateTopLevelNodes(node: ParseNode, level: Int, topLevelTokens: Int) {
```
Doesn't make a significant difference but you are right. This can be done with `Boolean`.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2254,19 +2254,19 @@ class SqlParser(private val ion: IonSystem) : Parser {
-     * Validates tree to make sure that the top level tokens are not found below the top level
+     * Validates tree to make sure that the top level tokens are not found below the top level.
+     * Top level tokens are the tokens or keywords which are valid to be used only at the top level in the query.
+     * i.e. these tokens cannot be used with a mix of other commands. Hence if more than one top level tokens are found
```
nit: "i.e. these tokens cannot be used with a mix of other ~commands~ top level tokens"? "commands" seems too generalized.

alancai98 commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -2254,19 +2254,19 @@ class SqlParser(private val ion: IonSystem) : Parser {
+     * i.e. these tokens cannot be used with a mix of other commands. Hence if more than one top level tokens are found
+     * in the query then it is invalid.
+     * [level] is the current traversal level in the parse tree.
+     * If [topLevelTokenSeen] is true, it means it has been encountered at least once before while traversing the parse tree.
```
nit: description of `topLevelTokenSeen` could be more more specific. Consider "...it means a top level token has been previously encountered while traversing the parse tree"

abhikuhikar replied:
I believe the variable name itself tells it all.

alancai98 reviewed (approved):
Couple very minor nits for the kdoc of `validateTopLevelNodes`. Rest lgtm
